org.eclipse.lsp4e/src/org/eclipse/lsp4e/DocumentContentSynchronizer.java

```diff
@@ -130,20 +136,20 @@ public DocumentContentSynchronizer(@NonNull LanguageServerWrapper languageServer
  	 * @param fn Asynchronous computation on the language server
  	 * @return Asynchronous result object.
  	 */
-	private <U> @NonNull CompletableFuture<U> executeOnCurrentVersionAsync(
+	<U> @NonNull CompletableFuture<VersionedResult<U>> executeOnCurrentVersionAsync(
 			Function<LanguageServer, ? extends CompletionStage<U>> fn) {
-		AtomicReference<CompletableFuture<U>> resValueFuture = new AtomicReference<>();
-		lastChangeFuture.updateAndGet(f -> {
-			CompletableFuture<U> valueFuture = f.thenComposeAsync(fn);
-			resValueFuture.set(valueFuture);
+		synchronized (this.guard) {
+			final long stampAtTimeExecuted = this.documentModificationStamp;
+			CompletableFuture<VersionedResult<U>> valueFuture = this.lastChangeFuture.
+					thenComposeAsync(fn).thenApply(u -> new VersionedResult<U>(u, stampAtTimeExecuted));
 			// We ignore any exceptions that happen when executing the given future
-			return valueFuture.handle((value, error) -> this.languageServer);
-		});
-		return resValueFuture.get();
+			this.lastChangeFuture = valueFuture.handle((value, error) -> this.languageServer);
+			return valueFuture;
+		}
 	}
 
-	CompletableFuture<LanguageServer> lastChangeFuture() {
-		return lastChangeFuture.get();
+	CompletableFuture<LanguageServer> didOpenFuture() {
```

Comment:
Do we expose immutable fields anywhere else? I tend to fall into old habits of using getters other than for classes that are `record`-style immutable POJOs but that's just superstition I guess!

More importantly, do we want this to be `public` rather than package private?

Reply:
> Do we expose immutable fields anywhere else?

Yes. See for example how LanguageServerWrapper just has a public final field for the ServerDefinition and how this server definition also has final fields.

> More importantly, do we want this to be public rather than package private?

The most restricted, the better, so package would be ok.

Reply:
OK, I can change to a package-level final field.